src/modules/kvs/kvs.c

```diff
@@ -1992,6 +1992,132 @@ static void stats_clear_request_cb (flux_t *h, flux_msg_handler_t *mh,
+{
+    struct kvsroot *root;
+    json_t *rootdir = NULL;
+    blobref_t ref;
+    void *data = NULL;
+    int len;
+
+    /* Namespace already exists */
+    if (zhash_lookup (ctx->roothash, namespace)) {
+        errno = EEXIST;
+        goto cleanup;
+    }
+
+    if (!(root = create_root (ctx, namespace, flags))) {
+        flux_log_error (ctx->h, "%s: create_root", __FUNCTION__);
+        goto cleanup;
+    }
+
+    if (!(rootdir = treeobj_create_dir ())) {
+        flux_log_error (ctx->h, "%s: treeobj_create_dir", __FUNCTION__);
+        goto cleanup_remove_root;
+    }
+
+    if (!(data = treeobj_encode (rootdir))) {
+        flux_log_error (ctx->h, "%s: treeobj_encode", __FUNCTION__);
+        goto cleanup_remove_root;
+    }
+    len = strlen (data);
+
+    if (blobref_hash (ctx->hash_name, data, len, ref) < 0) {
+        flux_log_error (ctx->h, "%s: blobref_hash", __FUNCTION__);
+        goto cleanup_remove_root;
+    }
+
+    setroot (ctx, root, ref, 0);
+
+    if (event_subscribe (ctx, namespace) < 0) {
+        flux_log_error (ctx->h, "%s: event_subscribe", __FUNCTION__);
+        goto cleanup_remove_root;
+    }
+
+    return 0;
+
+cleanup_remove_root:
+    remove_root (ctx, namespace);
+cleanup:
+    free (data);
+    json_decref (rootdir);
+    return -1;
+}
+
+static void namespace_relaycreate_request_cb (flux_t *h, flux_msg_handler_t *mh,
+                                              const flux_msg_t *msg, void *arg)
+{
+    kvs_ctx_t *ctx = arg;
+    const char *namespace;
+    int flags;
+
+    assert (ctx->rank == 0);
+
+    if (flux_request_unpack (msg, NULL, "{ s:s s:i }",
+                             "namespace", &namespace,
+                             "flags", &flags) < 0) {
+        flux_log_error (h, "%s: flux_request_unpack", __FUNCTION__);
+        goto error;
+    }
+
+    if (namespace_create (ctx, namespace, flags) < 0) {
+        flux_log_error (h, "%s: namespace_create", __FUNCTION__);
+        goto error;
+    }
+
+    errno = 0;
+error:
+    if (flux_respond (h, msg, errno, NULL) < 0)
+        flux_log_error (h, "%s: flux_respond", __FUNCTION__);
+}
+
+static void namespace_create_request_cb (flux_t *h, flux_msg_handler_t *mh,
+                                         const flux_msg_t *msg, void *arg)
+{
+    kvs_ctx_t *ctx = arg;
+    const char *namespace;
+    int flags;
+
+    if (flux_request_unpack (msg, NULL, "{ s:s s:i }",
+                             "namespace", &namespace,
+                             "flags", &flags) < 0) {
+        flux_log_error (h, "%s: flux_request_unpack", __FUNCTION__);
+        goto error;
+    }
+
+    if (ctx->rank == 0) {
+        if (namespace_create (ctx, namespace, flags) < 0) {
+            flux_log_error (h, "%s: namespace_create", __FUNCTION__);
+            goto error;
+        }
+    }
+    else {
+        flux_future_t *f;
+
+        /* XXX: future make asynchronous */
+        if (!(f = flux_rpc_pack (h, "kvs.namespace.relaycreate",
```

Comment:
Actually, it may not be desirable to use this "relay" strategy for namespace creation.  For example, if a job only spans a couple of nodes, it is not desirable for nodes that aren't using the namespace but are on the TBON path to the root to subscribe to its setroot events.

I would suggest changing the client RPC to directly communicate wtih rank 0 and eliminate the relay.

Reply:
sounds like a good idea